lib/iris/coords.py

```diff
@@ -2775,6 +2779,366 @@ def xml_element(self, doc):
+            for fastest varying index being either first or last). E.g. for
+            ``face_node_connectivity``, where the faces are 3-sided:
+            ``indices.shape[element_dim] = 3``.
+
+        """
+        # Configure the metadata manager.
+        self._metadata_manager = metadata_manager_factory(ConnectivityMetadata)
+
+        self._validate_arg_vs_list("start_index", start_index, [0, 1])
+        # indices array will be 2-dimensional, so must be either 0 or 1.
+        self._validate_arg_vs_list("element_dim", element_dim, [0, 1])
+        # TODO: find a better 'common' location to store valid cf_roles.
+        valid_cf_roles = [
+            "edge_node_connectivity",
+            "face_node_connectivity",
+            "face_edge_connectivity",
+            "face_face_connectivity",
+            "edge_face_connectivity",
+            "boundary_node_connectivity",
+            "volume_node_connectivity",
+            "volume_edge_connectivity",
+            "volume_face_connectivity",
+            "volume_volume_connectivity",
+        ]
+        self._validate_arg_vs_list("cf_role", cf_role, valid_cf_roles)
+
+        self._metadata_manager.start_index = start_index
+        self._metadata_manager.element_dim = element_dim
+        self._metadata_manager.cf_role = cf_role
+
+        super().__init__(
+            values=indices,
+            standard_name=None,
+            long_name=long_name,
+            var_name=var_name,
+            units="1",
+            attributes=attributes,
+        )
+
+    @staticmethod
+    def _validate_arg_vs_list(arg_name, arg, list):
+        if arg not in list:
+            error_msg = (
+                f"Invalid {arg_name} . Got: {arg} . Must be one of: "
+                f"{list} ."
+            )
+            raise ValueError(error_msg)
+
+    @staticmethod
+    def _lazy_element_lengths(array, element_dim):
+        element_mask_counts = da.sum(
+            da.ma.getmaskarray(array), axis=element_dim
+        )
+        max_element_size = array.shape[element_dim]
+        return max_element_size - element_mask_counts
+
+    @property
+    def _values(self):
+        # Overridden just to allow .setter override.
+        return super()._values
+
+    @_values.setter
+    def _values(self, values):
+        self._validate_indices(values)
+        # The recommended way of using the setter in super().
+        super(Connectivity, self.__class__)._values.fset(self, values)
+
+    @property
+    def cf_role(self):
+        """
+        The category of topological relationship that this connectivity
+        describes.
+        **Read-only** - validity of :attr:`indices` is dependent on
+        :attr:`cf_role`. A new :class:`Connectivity` must therefore be defined
+        if a different :attr:`cf_role` is needed.
+
+        """
+        return self._metadata_manager.cf_role
+
+    @property
+    def cf_role_element(self):
```

Comment:
The term `element` is kinda generic and overloaded... could be borrow from UGRID terminology itself?

It's almost akin to the `location` of a UGRID NetCDF data variable, but not for the focus of where data payload is on the mesh, but rather the focus is for what fundamentally is this connectivity relationship about e.g., connecting `𝛼` to `𝛽`, or `face` to `node`, where the focus/`location` is `face`.

Would you object to this being called `location`?

So you could say:
```python
connectivity.location
```
which seems more meaningful than:
```python
connectivity.cf_role_element
```
which is harder (for me) to determine what it's telling me.

Also, does this need to be public? ... not that it greatly matters, as all `Connectivity` will be managed by the `Mesh`... I think I've answered my own question... public is fine.

Reply:
Great idea since, as you say, we will be using `location` elsewhere to refer to these same things. "Located on this location" seems nice and self-reinforcing vs "Located on this element".

Reply:
In fact, the `cf_role_element` is fixed given the `cf_role`, so it's safe to calculate this and cache it once on the instance within the `__init__` rather than calculating it every call.

Reply:
Or, given a `cf_role = "face_node_connectivity"`, we could go with `connectivity.src` and `connectivity.tgt`, where:

- `connectivity.src` returns `"face"`
- `connectivity.tgt` returns `"node"`

This is building on the `"face"` being the source (`src`) of the connectivity, and `"node"` being the target (`tgt`) of the connectivity relationship.

There are few options here... but his might be even clearer.

Reply:
I really like this 👍

Reply:
Or even, e.g.,

- `connectivity.src_location` returns `"face"`
- `connectivity.src_dim` returns, say `0`
- `connectivity.tgt_location` returns `"node"`
- `connectivity.tgt_dim` returns `1`

Reply:
In this situation we would only have `src_dim` as a parameter of `Connectivity`. The class itself can maintain an awareness/understanding of the alternate values of `src_dim` and `tgt_dim`, which could be returned / used internally whenever needed.

Reply:
6139e563acdd0805e72a4a3e2ffa4391c3e07337